File: Code/Components/DanpheEMR.ServerModel/AccountingModels/LedgerModel.cs

using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel
{
    public class LedgerModel
    {
        [Key]
        public int LedgerId { get; set; }
        public int LedgerGroupId { get; set; }                        
        public string LedgerName { get; set; }                
        public int? LedgerReferenceId { get; set; }
        public string Description { get; set; }
        public int? SectionId { get; set; }
        public DateTime? CreatedOn { get; set; }
        public int CreatedBy { get; set; }
        public bool IsActive { get; set; }
        public bool? IsCostCenterApplicable { get; set; }
        public double? OpeningBalance { get; set; }
        public bool? DrCr { get; set; }
        public string Name { get; set; }
        public string LedgerType { get; set; }
        [NotMapped]
        public string PrimaryGroup { get; set; }
        [NotMapped]
        public string COA { get; set; }
        [NotMapped]
        public string LedgerGroupName { get; set; }
        public string Code { get; set; }
        public string PANNo { get; set; }
        public string Address { get; set; }
        public string MobileNo { get; set; }
	    public int? CreditPeriod { get; set; }
	    public decimal? TDSPercent { get; set; }
        public string LandlineNo { get; set; }

        [NotMapped]
        public int? EmployeeId { get; set; }
        [NotMapped]
        public string EmployeeName { get; set; }      
        [NotMapped]
        public string DepartmentName { get; set; }
        [NotMapped]
        public string SupplierName { get; set; }
        [NotMapped]
        public int? SupplierId { get; set; }
        [NotMapped]
        public string VendorName { get; set; }
        [NotMapped]
        public int? VendorId { get; set; }
        [NotMapped]
        public string SubCategoryName { get; set; }
        [NotMapped]
        public int? SubCategoryId { get; set; }
        [NotMapped]
        public string OrganizationName { get; set; }
        [NotMapped]
        public int? OrganizationId { get; set; }
        [NotMapped]
        public string ServiceDepartmentName { get; set; }
        [NotMapped]
        public int? ServiceDepartmentId { get; set; }
        [NotMapped]
        public int? ItemId { get; set; }
        [NotMapped]
        public string ItemName { get; set; }
        //sud-nagesh: 20June'20-- for Acc-Hospital Separation.
        public int? HospitalId { get; set; }

        [NotMapped]
        public bool? IsMapLedger { get; set; }

    }
    
}

File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/accounting/settings/shared/ledger.model.ts
import {
  NgForm,
  FormGroup,
  FormControl,
  Validators,
  FormBuilder,
  ReactiveFormsModule
} from '@angular/forms'

export class LedgerModel {
  public IsSelected: any;
  public LedgerId: number = 0;
  public LedgerName: string = null;
  public Code: string = null;
  public Type: string = null;
  public COA: string = null;
  public Description: string = null;
  public LedgerGroupId: number = null;
  public LedgerReferenceId: number = null;
  public OpeningBalance: number = 0;
  public DrCr: boolean = true;
  public CurrentBalance: number = 0;
  public CreatedBy: number = 0;
  public CreatedOn: string = null;
  public IsActive: boolean = true;
  public LedgerType: string = "";
  public Dr: boolean;
  public Cr: boolean;
  //public IsCostCenterApplicable: boolean = false;
  public LedgerGroupName: string = "";

  public checkSelectedLedger: boolean = false;
  public ChartOfAccountName: string = "";
  public PrimaryGroup: string = "";
  public SectionId: number = null;
  public Name: string = "";
  public ClosingBalance: number = 0;
  public ClosingBalwithDrCr: string = "";
  public IsCostCenterApplicable: boolean = null;
  public LedgerValidator: FormGroup = null;
  public PANNo: string = "";
  public Address: string = "";
  public MobileNo: string = "";
  public CreditPeriod: number = null;
  public TDSPercent: number = null;
  public LandlineNo: string = "";


  public EmployeeName :string = "";
  public EmployeeId:number = null;
  public DepartmentName : string = "";

  public SupplierName :string = "";
  public SupplierId:number = null;

  
  public VendorName :string = "";
  public VendorId:number = null;

  
  public SubCategoryName :string = "";
  public SubCategoryId:number = null;

  public OrganizationName :string = "";
  public OrganizationId:number = null;
  public ServiceDepartmentName :string = "";
  public ServiceDepartmentId:number = null;
  public ItemId:number = null;
  public ItemName :string = "";
  public IsMapLedger: boolean;
  public COAId:number=null;
  public selectedLedger:any=null;
  constructor() {

    var _formBuilder = new FormBuilder();
    this.LedgerValidator = _formBuilder.group({
      'LedgerName': ['', Validators.compose([Validators.required, Validators.maxLength(200)])],
      'LedgerGroupName': ['', Validators.compose([Validators.required])],
      'PrimaryGroup': ['', Validators.compose([Validators.required])],
      'COA': ['', Validators.compose([Validators.required])],
      'Dr': ['', Validators.compose([])],//conditional validation for DrCr(Dynamically change when required)
      'Cr': ['', Validators.compose([])]//conditional validation for DrCr(Dynamically change when required)
    });
  }

  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.LedgerValidator.dirty;
    else
      return this.LedgerValidator.controls[fieldName].dirty;
  }

  public IsValid(): boolean { if (this.LedgerValidator.valid) { return true; } else { return false; } }
  public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined) {
      return this.LedgerValidator.valid;
    }
    else
      return !(this.LedgerValidator.hasError(validator, fieldName));
  }
  //this function will update validator of required fields
  //Dynamically add validator
  public UpdateValidator(onOff: string, formControlName: string, validatorType: string) {
    let validator = null;
    if (validatorType == 'required' && onOff == "on") {
      validator = Validators.compose([Validators.required]);
    }
    else {
      validator = Validators.compose([]);
    }
    if (formControlName == "Dr") {
      this.LedgerValidator.controls['Dr'].validator = validator;
      this.LedgerValidator.controls['Dr'].updateValueAndValidity();
    }
    if (formControlName == "Cr") {
      this.LedgerValidator.controls['Cr'].validator = validator;
      this.LedgerValidator.controls['Cr'].updateValueAndValidity();
    }

  }
}
